lib/dao.js

```diff
@@ -2421,6 +2421,108 @@ DataAccessObject.updateAll = function(where, data, options, cb) {
   return cb.promise;
 };
 
+DataAccessObject.patchById = function patchById(id, data, options, cb) {
+  var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+  if (connectionPromise) {
+    return connectionPromise;
+  }
+  assert(arguments.length >= 3, 'At least three arguments are required');
+
+  var Model = this;
+  var where = {};
+  var idPropertyName = idName(Model);
+
+  where[idPropertyName] = id;
+  if (cb === undefined) {
+    if (typeof options === 'function') {
+      // patchById(id, data, cb)
+      cb = options;
+      options = {};
+    }
+  }
+
+  data = data || {};
+  cb = cb || utils.createPromiseCallback();
+  options = options || {};
+
+  assert((typeof data === 'object') && (data !== null),
+    'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+
+  var connector = Model.getDataSource().connector;
+  assert(typeof connector.update === 'function',
+    'update() must be implemented by the connector');
+
+  var hookState = {};
+  this.applyProperties(data);
+
+  var context = {
+    Model: Model,
+    instanceId: id,
```

Comment:
I am proposing to use the same context as we have for `updateAll`:

- Model
- where
- data
- hookState
- options

Note that instance id can be extracted from the `where` clause, see https://github.com/strongloop/loopback/blob/4cd84dcd8af5f4f90ff62f506a6bf49e6fc2036b/lib/persisted-model.js#L1508-L1518.

We can eventually put `getIdFromWhereByModelId` into a public LoopBack API if there is enough demand.

This is the reason why I am reluctant to add `instanceId`: when we have both `where` and `instanceId`, and a hook changes one of these properties only, it's not clear which one takes precedence. I suppose we could make `instanceId` read-only, although the downside is that `Object.defineProperty` incurs performance penalty :(